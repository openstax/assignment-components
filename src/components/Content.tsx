import React from "react";

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
type ComponentType = keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;

export interface ContentProps<T extends ComponentType | undefined> {
  className?: string;
  component?: T extends undefined ? undefined :
    T extends ComponentType ? React.ReactComponentElement<T>:
    never;
  html: string;
  block?: boolean;
}

export const Content = (<T extends ComponentType | undefined>(
  {html, component, block = false, ...props}: ContentProps<T>
) => {
  if (component !== undefined) {
    return React.cloneElement(component, {html, ...props});
  }
  if (block) {
    return <div dangerouslySetInnerHTML={{ __html: html }} {...props} />;
  } else {
    return <span dangerouslySetInnerHTML={{ __html: html }} {...props} />;
  }
});
