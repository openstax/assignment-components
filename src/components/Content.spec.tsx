import { Content } from './Content';
import renderer from 'react-test-renderer';

describe('Content', () => {
  const html = '<b>Content with HTML</b>';
  const CustomRenderer = ({ html = '' }: { html?: string }) => (
    <div className="custom-renderer" dangerouslySetInnerHTML={{ __html: html }} />
  );

  it('matches snapshot', () => {
     const tree = renderer.create(
       <Content html={html} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with a custom renderer if set', () => {
    const tree = renderer.create(
      <Content html={html} component={<CustomRenderer />} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
