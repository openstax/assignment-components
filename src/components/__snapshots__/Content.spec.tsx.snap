// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content matches snapshot 1`] = `
<span
  dangerouslySetInnerHTML={
    Object {
      "__html": "<b>Content with HTML</b>",
    }
  }
/>
`;

exports[`Content renders with a custom renderer if set 1`] = `
<div
  className="custom-renderer"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<b>Content with HTML</b>",
    }
  }
/>
`;
